clients/githubrepo/branches.go

```diff
@@ -31,24 +31,70 @@ const (
 	refPrefix     = "refs/heads/"
 )
 
-type refUpdateRule struct {
+// See https://github.community/t/graphql-api-protected-branch/14380
+/* Example of query:
+	query {
+  repository(owner: "laurentsimon", name: "test3") {
+    branchProtectionRules(first: 100) {
+      nodes {
+        pushAllowances (first:10){
+          edges {
+            node {
+              id
+            }
+          }
+        }
+        allowsDeletions
+        allowsForcePushes
+        dismissesStaleReviews
+        isAdminEnforced
+        requiresApprovingReviews
+        requiredApprovingReviewCount
+        requiresStatusChecks
+        requiresStrictStatusChecks
+        restrictsPushes
+        branchProtectionRuleConflicts(first:3) {
+          __typename
+          nodes {
+            __typename
+          }
+        }
+        pattern
+        matchingRefs(first: 100) {
+          nodes {
+            name
+
+          }
+        }
+      }
+    }
+	// I don't think `refs` is needed. The query qorks without it.
+    refs(first: 100, refPrefix: "refs/heads/") {
+      nodes {
+        name
+        __typename
+      }
+    }
+  }
+}
+*/
+type branchProtectionRule struct {
+	DismissesStaleReviews        *bool
+	IsAdminEnforced              *bool
+	RequiresStrictStatusChecks   *bool
+	RequiresStatusChecks         *bool
 	AllowsDeletions              *bool
 	AllowsForcePushes            *bool
 	RequiredApprovingReviewCount *int32
 	RequiresCodeOwnerReviews     *bool
 	RequiresLinearHistory        *bool
 	RequiredStatusCheckContexts  []string
-}
-
-type branchProtectionRule struct {
-	DismissesStaleReviews      *bool
-	IsAdminEnforced            *bool
-	RequiresStrictStatusChecks *bool
+	// TODO: verify there is no conflicts.
+	// BranchProtectionRuleConflicts interface{}
 }
 
 type branch struct {
 	Name                 *string
-	RefUpdateRule        *refUpdateRule
```

Comment:
`RefUpdateRule` allows us to access settings viewable to non-admins. Won't removing this mean going back to our previous state where `Branch-Protection` check is mostly useful for admins?

![image](https://user-images.githubusercontent.com/4825078/141655214-e43f3b0b-93ec-4196-901e-b3ad6ecb09e6.png)

Reply:
good catch. I'm honestly confused. Not using it is what seemed to fix the problem. Can you try to reproduce the results I got with this PR? I actually tested in with a non-admin token all along. Maybe what they mean in their doc is `rules that are viewable *only* by non-admins`.
I will aslo try to verify the findings once more, and let's sync wafter that.

Reply:
I actually haven't been able to reproduce the problem from my side so far. But here's my preliminary suggestion - let's keep `RefUpdateRule` *and* the changes you have made as of now. We can fallback to the values returned by `RefUpdatedRule` if `BranchProtectionRule` returns `nil`. Do you think that would fix the problem while maintaining status quo?

Reply:
I suppose that'd work. I'm only worried that the code I added will have unwanted side effects. I'll do another round of  testing. In the graphQl API, they have `branchProtectionRules : A list of branch protection rules for this repository.` with no limitation whatsoever on differences between admin/non-admin, which is really confusing: I'll probably create a GH issue and ask them directly.

BTW, where did you find the list of branch protection settings that require admin tokens?

Reply:
My assumption was that since `RefUpdateRules` has all non-admin rules, `BranchProtectionRules` must be a super set of all rules available to both non-admins and admins.

Reply:
I think that's correct. It just seems that when running as admin, `RefUpdateRules` is not available at all and everything appears in `BranchProtectionRules` instead.